File: python/src/drive.py
import os
import json
from pathlib import Path
from google.oauth2 import service_account
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload

# Google Drive APIのスコープ
SCOPES = ['https://www.googleapis.com/auth/drive.file']

async def authorize(logger):
    """
    Google Drive APIの認証を行う
    
    Args:
        logger: ロガー
        
    Returns:
        google.auth.credentials.Credentials: 認証済みのクレデンシャル
    """
    try:
        # 環境変数から認証情報のパスを取得
        credentials_path = os.getenv('GOOGLE_APPLICATION_CREDENTIALS') or \
                          str(Path(__file__).parent.parent / 'credentials' / 'google-credentials.json')
        
        # 認証情報ファイルが存在するか確認
        if not os.path.exists(credentials_path):
            raise FileNotFoundError(f'認証情報ファイルが見つかりません: {credentials_path}')
        
        # 認証情報を読み込む
        with open(credentials_path, 'r') as f:
            credentials_data = json.load(f)
        
        # サービスアカウント認証の場合
        if 'type' in credentials_data and credentials_data['type'] == 'service_account':
            credentials = service_account.Credentials.from_service_account_file(
                credentials_path, scopes=SCOPES)
            return credentials
        
        # OAuth2認証の場合
        elif 'installed' in credentials_data or 'web' in credentials_data:
            # トークンファイルのパス
            token_path = str(Path(__file__).parent.parent / 'credentials' / 'token.json')
            
            credentials = None
            # トークンファイルが存在する場合は読み込む
            if os.path.exists(token_path):
                credentials = Credentials.from_authorized_user_info(
                    json.loads(open(token_path, 'r').read()), SCOPES)
            
            # 有効な認証情報がない場合は新たに取得
            if not credentials or not credentials.valid:
                if credentials and credentials.expired and credentials.refresh_token:
                    credentials.refresh(Request())
                else:
                    flow = InstalledAppFlow.from_client_secrets_file(credentials_path, SCOPES)
                    credentials = flow.run_local_server(port=0)
                
                # トークンを保存
                with open(token_path, 'w') as token:
                    token.write(credentials.to_json())
            
            return credentials
        else:
            raise ValueError('サポートされていない認証タイプです。')
    
    except Exception as e:
        logger.error(f'Google認証中にエラーが発生しました: {str(e)}', exc_info=True)
        raise

async def upload_to_drive(file_path, file_name, channel_name, logger):
    """
    ファイルをGoogle Driveにアップロードする
    
    Args:
        file_path (str): アップロードするファイルのパス
        file_name (str): ファイル名
        channel_name (str): 音声チャンネル名（フォルダ名として使用）
        logger: ロガー
        
    Returns:
        dict: アップロード結果
    """
    try:
        # 設定ファイルの読み込み
        with open('config.json', 'r', encoding='utf-8') as f:
            config = json.load(f)
        
        # Google認証
        credentials = await authorize(logger)
        drive_service = build('drive', 'v3', credentials=credentials)
        
        # ファイルのメタデータ
        file_metadata = {
            'name': file_name,
            'description': f'Discord音声チャンネル「{channel_name}」の録音 - {datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")}'
        }
        
        # 保存先フォルダが指定されている場合
        if config.get('googleDriveFolderId'):
            file_metadata['parents'] = [config['googleDriveFolderId']]
        
        # メディアデータ
        mime_type = get_mime_type(file_name)
        media = MediaFileUpload(file_path, mimetype=mime_type, resumable=True)
        
        # ファイルのアップロード
        logger.info(f'Google Driveにアップロード中: {file_name}')
        file = drive_service.files().create(
            body=file_metadata,
            media_body=media,
            fields='id,name,webViewLink'
        ).execute()
        
        logger.info(f'アップロード完了: {file.get("name")} (ID: {file.get("id")})')
        
        # ファイルの共有設定を変更（リンクを知っている全員が閲覧可能に）
        drive_service.permissions().create(
            fileId=file.get('id'),
            body={
                'role': 'reader',
                'type': 'anyone'
            }
        ).execute()
        
        logger.info(f'共有設定を変更しました: {file.get("id")}')
        
        return {
            'id': file.get('id'),
            'name': file.get('name'),
            'web_view_link': file.get('webViewLink')
        }
    
    except Exception as e:
        logger.error(f'Google Driveへのアップロード中にエラーが発生しました: {str(e)}', exc_info=True)
        raise

def get_mime_type(file_name):
    """
    ファイル拡張子からMIMEタイプを取得
    
    Args:
        file_name (str): ファイル名
        
    Returns:
        str: MIMEタイプ
    """
    ext = os.path.splitext(file_name)[1].lower()
    
    mime_types = {
        '.mp3': 'audio/mpeg',
        '.wav': 'audio/wav',
        '.ogg': 'audio/ogg',
        '.m4a': 'audio/mp4',
        '.flac': 'audio/flac'
    }
    
    return mime_types.get(ext, 'application/octet-stream')

# datetime モジュールのインポート
import datetime

File: src/drive.js
const fs = require('fs');
const path = require('path');
const { google } = require('googleapis');
const config = require('../config.json');

// Google Drive APIのスコープ
const SCOPES = ['https://www.googleapis.com/auth/drive.file'];

/**
 * Google Drive APIの認証を行う
 * @param {Logger} logger - ロガー
 * @returns {google.auth.OAuth2} - 認証済みのOAuth2クライアント
 */
async function authorize(logger) {
  try {
    // 環境変数から認証情報のパスを取得
    const credentialsPath = process.env.GOOGLE_APPLICATION_CREDENTIALS || 
                           path.join(__dirname, '../credentials/google-credentials.json');
    
    // 認証情報ファイルが存在するか確認
    if (!fs.existsSync(credentialsPath)) {
      throw new Error(`認証情報ファイルが見つかりません: ${credentialsPath}`);
    }
    
    // 認証情報を読み込む
    const credentials = JSON.parse(fs.readFileSync(credentialsPath, 'utf8'));
    
    // サービスアカウント認証の場合
    if (credentials.type === 'service_account') {
      const auth = new google.auth.GoogleAuth({
        keyFile: credentialsPath,
        scopes: SCOPES
      });
      return auth;
    } 
    // OAuth2認証の場合
    else if (credentials.installed || credentials.web) {
      const { client_secret, client_id, redirect_uris } = credentials.installed || credentials.web;
      const oauth2Client = new google.auth.OAuth2(client_id, client_secret, redirect_uris[0]);
      
      // トークンファイルのパス
      const tokenPath = path.join(__dirname, '../credentials/token.json');
      
      // トークンファイルが存在する場合は読み込む
      if (fs.existsSync(tokenPath)) {
        const token = JSON.parse(fs.readFileSync(tokenPath, 'utf8'));
        oauth2Client.setCredentials(token);
        return oauth2Client;
      } else {
        throw new Error('OAuth2トークンが見つかりません。認証フローを実行してください。');
      }
    } else {
      throw new Error('サポートされていない認証タイプです。');
    }
  } catch (error) {
    logger.error(`Google認証中にエラーが発生しました: ${error.message}`);
    throw error;
  }
}

/**
 * ファイルをGoogle Driveにアップロードする
 * @param {string} filePath - アップロードするファイルのパス
 * @param {string} fileName - ファイル名
 * @param {string} channelName - 音声チャンネル名（フォルダ名として使用）
 * @param {Logger} logger - ロガー
 * @returns {Object} - アップロード結果
 */
async function uploadToDrive(filePath, fileName, channelName, logger) {
  try {
    // Google認証
    const auth = await authorize(logger);
    const drive = google.drive({ version: 'v3', auth });
    
    // ファイルのメタデータ
    const fileMetadata = {
      name: fileName,
      description: `Discord音声チャンネル「${channelName}」の録音 - ${new Date().toLocaleString()}`,
    };
    
    // 保存先フォルダが指定されている場合
    if (config.googleDriveFolderId) {
      fileMetadata.parents = [config.googleDriveFolderId];
    }
    
    // メディアデータ
    const media = {
      mimeType: getMimeType(fileName),
      body: fs.createReadStream(filePath)
    };
    
    // ファイルのアップロード
    logger.info(`Google Driveにアップロード中: ${fileName}`);
    const response = await drive.files.create({
      resource: fileMetadata,
      media: media,
      fields: 'id,name,webViewLink'
    });
    
    logger.info(`アップロード完了: ${response.data.name} (ID: ${response.data.id})`);
    
    // ファイルの共有設定を変更（リンクを知っている全員が閲覧可能に）
    await drive.permissions.create({
      fileId: response.data.id,
      requestBody: {
        role: 'reader',
        type: 'anyone'
      }
    });
    
    logger.info(`共有設定を変更しました: ${response.data.id}`);
    
    return response.data;
  } catch (error) {
    logger.error(`Google Driveへのアップロード中にエラーが発生しました: ${error.message}`);
    throw error;
  }
}

/**
 * ファイル拡張子からMIMEタイプを取得
 * @param {string} fileName - ファイル名
 * @returns {string} - MIMEタイプ
 */
function getMimeType(fileName) {
  const ext = path.extname(fileName).toLowerCase();
  
  const mimeTypes = {
    '.mp3': 'audio/mpeg',
    '.wav': 'audio/wav',
    '.ogg': 'audio/ogg',
    '.m4a': 'audio/mp4',
    '.flac': 'audio/flac'
  };
  
  return mimeTypes[ext] || 'application/octet-stream';
}

module.exports = {
  uploadToDrive
};